src/waltz/ip/fd_netlink.c

```diff
@@ -588,136 +588,146 @@ fd_nl_load_arp_table( fd_nl_t *           nl,
+	/* function was interrupted - don't switch to new routing table */
 
-      return FD_IP_ERROR; /* return failure */
-    }
+	return FD_IP_RETRY; /* return transient failure */
+      }
 
-    struct ndmsg *  msg       = NLMSG_DATA( h );
-    struct rtattr * rat       = RTM_RTA(msg);
-    long            ratmsglen = (long)RTM_PAYLOAD( h );
-
-    /* store the interface */
-    entry->ifindex = (uint)msg->ndm_ifindex;
-    entry->flags  |= FD_NL_ARP_FLAGS_IFINDEX;
-
-    /* we want to skip some entries based on state
-       here are the states:
-           NUD_INCOMPLETE   a currently resolving cache entry
-           NUD_REACHABLE    a confirmed working cache entry
-           NUD_STALE        an expired cache entry
-           NUD_DELAY        an entry waiting for a timer
-           NUD_PROBE        a cache entry that is currently reprobed
-           NUD_FAILED       an invalid cache entry
-           NUD_NOARP        a device with no destination cache
-           NUD_PERMANENT    a static entry
-
-       Keeping:
-           NUD_REACHABLE    valid, so use it
-           NUD_STALE        probably better to use the existing address
-                              than wait or discard packet
-           NUD_DELAY        an entry waiting for a timer
-           NUD_PROBE        being reprobed, so it's probably valid
-           NUD_PERMANENT    a static entry, so use it */
-
-    while( RTA_OK( rat, ratmsglen ) ) {
-      uchar * rta_data    = RTA_DATA( rat );
-      ulong   rta_data_sz = RTA_PAYLOAD( rat );
-
-      switch( rat->rta_type ) {
-        case NDA_DST:
-          if( rta_data_sz != 4 ) {
-            /* we only support IPv4 */
-            entry->flags |= FD_NL_ARP_FLAGS_UNSUPPORTED;
-          } else {
-            uint dst_ip_addr;
-            memcpy( &dst_ip_addr, rta_data, 4 );
-            entry->dst_ip_addr = ntohl( dst_ip_addr );
-
-            entry->flags |= FD_NL_ARP_FLAGS_IP_ADDR;
-          }
-          break;
+      if( h->nlmsg_type == NLMSG_ERROR ) {
+	struct nlmsgerr * err = NLMSG_DATA(h);
 
-        case NDA_LLADDR:
-          if( FD_UNLIKELY( rta_data_sz != 6 ) ) {
-            /* we only support ethernet, but these entries are found */
-            /* on some machines, so silently skip them */
-            entry->flags |= FD_NL_ARP_FLAGS_UNSUPPORTED;
-          } else {
-            memcpy( &entry->mac_addr[0], rta_data, 6 );
+	FD_LOG_WARNING(( "netlink returned data with error: %d %s", -err->error, strerror( -err->error) ));
 
-            entry->flags |= FD_NL_ARP_FLAGS_MAC_ADDR;
-          }
-          break;
+	/* error occurred - don't switch to new routing table */
 
+	return FD_IP_ERROR; /* return failure */
       }
 
-      if( entry->flags & FD_NL_ARP_FLAGS_UNSUPPORTED ) {
-        break;
+      /* we're done if:
+       *        type is NLMSG_DONE
+       *    OR  nlmsg_flags does not contain NLM_F_MULTI */
+      done |= ( h->nlmsg_type == NLMSG_DONE ) | !( h->nlmsg_flags & NLM_F_MULTI );
```

Comment:
If `h->nlmsg_type == NLMSG_DONE`, should we break? I think the while loop below might have 0 iterations, but we'll still insert an entry.

Reply:
The flags NLMSG_DONE is set on a valid message, so we need to finish iterating thru the attributes.